src/lib.rs

```diff
@@ -572,67 +477,198 @@ impl KafkaJsonToDelta {
+        state.delta_writer.update_table().await?;
+
+        let mut attempt_number: u32 = 0;
+
+        loop {
+            state.delta_writer.update_table().await?;
+
+            if !self.are_partition_offsets_match(state) {
+                debug!("Transaction attempt failed. Got conflict partition offsets from delta store. Resetting consumer");
+                self.reset_state_guarded(state).await?;
+                // todo delete parquet file
+                return Ok(());
+            }
+
+            let version = state.delta_writer.table_version() + 1;
+            let actions = self.build_actions(&partition_offsets, add.clone()).await;
+            let commit_result = state.delta_writer.commit_version(version, actions).await;
+
+            match commit_result {
+                Ok(v) => {
+                    assert_eq!(v, version);
+
+                    state.delta_partition_offsets = partition_offsets;
+
+                    self.log_delta_write_completed(version, &delta_write_timer)
+                        .await;
+                    return Ok(());
                 }
-                _ => {
+                Err(e) => match e {
+                    DeltaTransactionError::VersionAlreadyExists { .. }
+                        if attempt_number > DEFAULT_DELTA_MAX_RETRY_COMMIT_ATTEMPTS + 1 =>
+                    {
+                        debug!("Transaction attempt failed. Attempts exhausted beyond max_retry_commit_attempts of {} so failing.", DEFAULT_DELTA_MAX_RETRY_COMMIT_ATTEMPTS);
+                        self.log_delta_write_failed().await;
+                        return Err(e.into());
+                    }
+                    DeltaTransactionError::VersionAlreadyExists { .. } => {
+                        attempt_number += 1;
+                        debug!("Transaction attempt failed. Incrementing attempt number to {} and retrying.", attempt_number);
+                    }
+                    _ => {
+                        self.log_delta_write_failed().await;
+                        return Err(e.into());
+                    }
+                },
+            }
+        }
+    }
+
+    /// Checks whether partition offsets from last writes matches the ones from delta log
+    /// If not then other consumers already processed them.
+    fn are_partition_offsets_match(&self, state: &mut ProcessingState) -> bool {
+        let mut result = true;
+        for (partition, offset) in &state.delta_partition_offsets {
+            let version = state
+                .delta_writer
+                .last_transaction_version(&self.app_id_for_partition(*partition));
+
+            if let Some(version) = version {
+                // if messages in kafka are consecutive then offset should always be `version+1` for
+                // safe commit, but since kafka does not guarantee contiguous offsets, we only check for `less than`.
+                if *offset != version {
```

Comment:
Another leftover, I was leaving these comments ion questionable places, but right now we only check for equality. Btw, since kafka might not be consecutive, I've introduced the `delta_partition_offsets` which actually a 1-to-1 match with latest known commit from delta app versions so we safe to check on `=` here